File: 백준/Gold/3079. 입국심사/입국심사.py
import sys
input = sys.stdin.readline


n, m = map(int, input().split())
T = []

for _ in range(n):
    T.append(int(input()))

left = min(T)
right = max(T) * m
answer = 1e20

while left <= right:
    mid = (left+right) // 2

    total = 0
    for t in T:
        total += mid // t

    if total >= m:
        answer = min(answer, mid)
        right = mid - 1
    else:
        left = mid + 1

print(answer)

File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;

class Solution {
    long answer = 1_000_000_000_000_000_000L;
        
    public boolean calculate(int n, long mid, int[] times) {
        long cnt = 0;
        long maxValue = 0;
        
        for (int i = 0; i < times.length; i++) {
            long tmp  = mid / times[i];
            maxValue = Math.max(maxValue, tmp * times[i]);
            cnt += tmp;
        }
        
        if (cnt >= n) {
            answer = Math.min(answer, maxValue);
            return true;
        }
        
        return false;
    }
    
    public long solution(int n, int[] times) {
        long left = 0;
        long right = 1_000_000_000_000_000_000L;
        
        Arrays.sort(times);
        
        while (left <= right) {
            long mid = (left + right) / 2;
            boolean flag = calculate(n, mid, times);
            
            if (flag) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        
        return answer;
    }
}